### ISSUE TITLE
padding='SAME'

### ISSUE THREAD

**Issue by koz4k** [ibab/tensorflow-wavenet#186]
Is there any particular reason why you use padding='SAME' instead of 'VALID' in all convolutional layers?
As I understand (please correct me if I'm wrong), it means that the network doesn't get the whole input for the first few samples in each batch, and the amount of samples with incomplete input is roughly equal to the size of the receptive field. It can cause the network to train poorly. This is a problem for me - I don't have so much memory on my graphics card, so I had to decrease sample_size to 10000 for the data to fit in the memory. I use 5 layer stacks with maximum dilation 1024, so my receptive field is 5 * 1024 * 2 = 10240 > 10000, which means that no sample gets the whole input and half of them gets less than half of the input. It will also be a problem for anyone who keeps sample_size = 100000 and increases the receptive field 10x. I'd like to fix this but don't want to break anything, that's why I'm asking.

**Comment by nakosung** [comment]
There is a PR related to this issue. https://github.com/ibab/tensorflow-wavenet/pull/118. :)

**Comment by jyegerlehner** [comment]
SAME vs VALID: If you use valid convolutions, each subsequent wavenet layer decreases the number of time steps at its output, whereas wavenet needs to preserve the number of time steps such that there is a predicted output at each time step that is present in the input. The target value provided to the cross entropy loss is identical to the input, only shifted one time step.

If you aren't specifying --window when you generate, you will get the default which is 8000. So any net with an input receptive field > 8000 will never have its receptive field filled.  So if you aren't specifying --window, you need to set it to 10240 (your receptive field size).

I think you also have a problem with sample_size = 10000 < receptive field size (10240). In training, the last 240 at the tail-end never get filled. I think you should either increase sample_size or decrease receptive field size.

If you are so memory-constrained, perhaps you'd be better off reducing sample rate from 16K (assuming you're dealing wit audio) rather than making sample_size so small.

**Comment by koz4k** [comment]
Ok, sorry for duplicating the issue.
I agree that VALID padding decreases the number of samples, I was thinking about prepending each batch with extra samples so that each "actual" sample can be calculated using the whole input, and then calculating cross entropy only using the "actual" samples. Then AudioReader should return overlapping batches, so that the extra samples come from the end of the previous batch. #118 doesn't do batch overlapping, which means that it always ignores part of the data. Implementing this would make the loss calculation more correct and reduce the memory requirements without ignoring data, at the cost of some complication of the code.
I think I'll do it for my own use, are you sure you wouldn't want it merged?

**Comment by jyegerlehner** [comment]
> I think I'll do it for my own use, are you sure you wouldn't want it merged?

No I wouldn't say that. It might be a great idea, I don't understand it yet. In any case, if you have something good please submit a PR and we'll see if it gets merged.

**Comment by koz4k** [comment]
Okay :)

**Comment by jyegerlehner** [comment]
@koz4k I've pondered this a bit more and think you are on to something.

I've always been opposed to "invalid" convolutions as they must necessarily inject noise when they operate on the padded input values. I don't know why it didn't occur to me to be more concerned about this before.

So you propose changing "SAME" to "VALID" [here](https://github.com/ibab/tensorflow-wavenet/blob/master/wavenet/ops.py#L54-L57) right? So the two things that come to mind that need to change as a part of this are 1) the slicing to match the output up with the shifted input when computing the cross entropy error, and 2) the AudioReader needs to proceed through the .wav file in increments equal to the output size (which is reduced relative to the input), which is smaller than the size it provides in each batch, in order to achieve the overlap you mentioned.

I'll be interested to see your changes!

**Comment by koz4k** [comment]
@jyegerlehner that's exactly what I have in mind.
